% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  pdftitle={Lab report},
  pdfauthor={Amir Rakhimov},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{Lab report}
\usepackage{etoolbox}
\makeatletter
\providecommand{\subtitle}[1]{% add subtitle to \maketitle
  \apptocmd{\@title}{\par {\large #1 \par}}{}{}
}
\makeatother
\subtitle{National Institute of Genetics, Biological networks lab}
\author{Amir Rakhimov}
\date{2022-11-15}

\begin{document}
\maketitle

\hypertarget{goals}{%
\section{Goals}\label{goals}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Reproduce results from \emph{Biagi et al.} (2017) as close as
  possible\\
\item
  Obtain taxonomic composition
\end{enumerate}

\hypertarget{methodology}{%
\section{Methodology}\label{methodology}}

\hypertarget{software}{%
\subsection{Software}\label{software}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Ubuntu 20.04 LTS (Windows subsystem for Linux)
\item
  conda v22.9.0
\item
  Qiime2 v2022.8-py38-linux-conda
\end{enumerate}

\hypertarget{experimental-workflow}{%
\subsection{Experimental workflow}\label{experimental-workflow}}

\hypertarget{import-pandaseq-demultiplexed-single-end-reads-as-singleendfastqmanifestphred33v2}{%
\subsubsection{\texorpdfstring{1. Import pandaseq demultiplexed
single-end reads as
\textbf{SingleEndFastqManifestPhred33V2}}{1. Import pandaseq demultiplexed single-end reads as SingleEndFastqManifestPhred33V2}}\label{import-pandaseq-demultiplexed-single-end-reads-as-singleendfastqmanifestphred33v2}}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{qiime}\NormalTok{ tools import }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}type}  \StringTok{\textquotesingle{}SampleData[SequencesWithQuality]\textquotesingle{}} \DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}input{-}path}\NormalTok{ filenames.txt }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}output{-}path}\NormalTok{ single{-}end{-}demux.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}input{-}format}\NormalTok{ SingleEndFastqManifestPhred33V2 }
  
\CommentTok{\# quality scores: 1 722 650 reads in total}
\ExtensionTok{qiime}\NormalTok{ demux summarize }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}data}\NormalTok{ single{-}end{-}demux.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}visualization}\NormalTok{ single{-}end{-}demux.qzv}
\CommentTok{\# Min: 48678; Max: 49499; Mean: 49218.571429}
\end{Highlighting}
\end{Shaded}

\hypertarget{denoise-with-dada2}{%
\subsubsection{2. Denoise with DADA2:}\label{denoise-with-dada2}}

\begin{itemize}
\tightlist
\item
  no truncation or trimming (trim on nt 1)
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{qiime}\NormalTok{ dada2 denoise{-}single }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}demultiplexed{-}seqs}\NormalTok{ single{-}end{-}demux.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}trim{-}left}\NormalTok{ 1 }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}trunc{-}len}\NormalTok{ 0 }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}representative{-}sequences}\NormalTok{ rep{-}seqs{-}dada2{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}table}\NormalTok{ table{-}dada2{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}denoising{-}stats}\NormalTok{ stats{-}dada2{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}n{-}threads}\NormalTok{ 8 }
  
\CommentTok{\#\#\#\#\#\#}
\CommentTok{\# visualize your dada2 stats}
\ExtensionTok{qiime}\NormalTok{ metadata tabulate }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}m{-}input{-}file}\NormalTok{ stats{-}dada2{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}visualization}\NormalTok{ stats{-}dada2{-}no{-}trunc.qzv}

\CommentTok{\# generate feature table (frequencies)}
\ExtensionTok{qiime}\NormalTok{ feature{-}table summarize }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}table}\NormalTok{ table{-}dada2{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}visualization}\NormalTok{ table{-}dada2{-}no{-}trunc.qzv}

\CommentTok{\# generate feature table (sequences)}
\ExtensionTok{qiime}\NormalTok{ feature{-}table tabulate{-}seqs }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}data}\NormalTok{ rep{-}seqs{-}dada2{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}visualization}\NormalTok{ rep{-}seqs{-}dada2{-}no{-}trunc.qzv }
\end{Highlighting}
\end{Shaded}

\hypertarget{cluster-into-97-identical-otus}{%
\subsubsection{3. Cluster into 97\% identical
OTUs}\label{cluster-into-97-identical-otus}}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{qiime}\NormalTok{ vsearch cluster{-}features{-}de{-}novo }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}table}\NormalTok{ table{-}dada2{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}sequences}\NormalTok{ rep{-}seqs{-}dada2{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}perc{-}identity}\NormalTok{ 0.97 }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}clustered{-}table}\NormalTok{ table{-}dn{-}97{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}clustered{-}sequences}\NormalTok{ rep{-}seqs{-}dn{-}97{-}no{-}trunc.qza}

\CommentTok{\#\# visualize clustering results}
\ExtensionTok{qiime}\NormalTok{ feature{-}table tabulate{-}seqs }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}data}\NormalTok{ rep{-}seqs{-}dn{-}97.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}visualization}\NormalTok{ rep{-}seqs{-}dn{-}97.qzv }
\end{Highlighting}
\end{Shaded}

\hypertarget{chimera-check-and-filter}{%
\subsubsection{4. Chimera check and
filter}\label{chimera-check-and-filter}}

\begin{itemize}
\tightlist
\item
  de novo check with uchime
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# chimera check: de novo with uchime}
\ExtensionTok{qiime}\NormalTok{ vsearch uchime{-}denovo }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}table}\NormalTok{ table{-}dn{-}97{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}sequences}\NormalTok{ rep{-}seqs{-}dn{-}97{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}output{-}dir}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc}
  
\CommentTok{\#\# visualize summary stats}
\ExtensionTok{qiime}\NormalTok{ metadata tabulate }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}m{-}input{-}file}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc/stats.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}visualization}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc/stats.qzv}
\end{Highlighting}
\end{Shaded}

\begin{itemize}
\tightlist
\item
  filter chimeras but retain borderline chimeras
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# chimera filtering: retain borderline}
\ExtensionTok{qiime}\NormalTok{ feature{-}table filter{-}features }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}table}\NormalTok{ table{-}dn{-}97{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}m{-}metadata{-}file}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc/chimeras.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}exclude{-}ids} \DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}filtered{-}table}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc/table{-}nonchimeric{-}w{-}borderline{-}97{-}no{-}trunc.qza}
  
\ExtensionTok{qiime}\NormalTok{ feature{-}table filter{-}seqs }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}data}\NormalTok{ rep{-}seqs{-}dn{-}97{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}m{-}metadata{-}file}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc/chimeras.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}exclude{-}ids} \DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}filtered{-}data}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc/rep{-}seqs{-}nonchimeric{-}w{-}borderline{-}97{-}no{-}trunc.qza}
  
\ExtensionTok{qiime}\NormalTok{ feature{-}table summarize }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}table}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc/table{-}nonchimeric{-}w{-}borderline{-}97{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}visualization}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc/table{-}nonchimeric{-}w{-}borderline{-}97{-}no{-}trunc.qzv}
\end{Highlighting}
\end{Shaded}

\hypertarget{taxonomic-classification}{%
\subsubsection{5. Taxonomic
classification}\label{taxonomic-classification}}

\begin{itemize}
\tightlist
\item
  Use Greengenes 13\_5 database with 97\% identical OTUs
\item
  Import FASTA sequences of OTUs as a QIIME artifact
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# import otus}
\ExtensionTok{qiime}\NormalTok{ tools import }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}type} \StringTok{\textquotesingle{}FeatureData[Sequence]\textquotesingle{}} \DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}input{-}path}\NormalTok{ ./gg\_13\_5\_otus/rep\_set/97\_otus.fasta }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}output{-}path}\NormalTok{ 97\_otus.qza}
\end{Highlighting}
\end{Shaded}

\begin{itemize}
\tightlist
\item
  Import taxonomical information as HeaderlessTSVTaxonomyFormat
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# import taxonomy}
\ExtensionTok{qiime}\NormalTok{ tools import }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}type} \StringTok{\textquotesingle{}FeatureData[Taxonomy]\textquotesingle{}} \DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}input{-}format}\NormalTok{ HeaderlessTSVTaxonomyFormat }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}input{-}path}\NormalTok{ ./gg\_13\_5\_otus/taxonomy/97\_otu\_taxonomy.txt }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}output{-}path}\NormalTok{ ref{-}taxonomy.qza}
\end{Highlighting}
\end{Shaded}

\begin{itemize}
\tightlist
\item
  Extract reference reads for classification : no truncation, no
  minimal/maximal length filtering
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# extract reference reads: no truncation, no min or max length filtering}
\ExtensionTok{qiime}\NormalTok{ feature{-}classifier extract{-}reads }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}sequences}\NormalTok{ 97\_otus.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}f{-}primer}\NormalTok{ CCTACGGGNGGCWGCAG }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}r{-}primer}\NormalTok{ GACTACHVGGGTATCTAATCC }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}trunc{-}len} \AttributeTok{{-}1} \DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}min{-}length}\NormalTok{ 0 }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}p{-}max{-}length}\NormalTok{ 0 }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}reads}\NormalTok{ ref{-}seqs.qza}
\end{Highlighting}
\end{Shaded}

Primers: S-D-Bact-0341-b-S-17 (5'-CCTACGGGNGGCWGCAG-3') and
S-D-Bact-0785-a-A-21 (5'-GACTACHVGGGTATCTAATCC-3')

\begin{itemize}
\tightlist
\item
  Train the Naive Bayes classifier
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# train the classifier}
\ExtensionTok{qiime}\NormalTok{ feature{-}classifier fit{-}classifier{-}naive{-}bayes }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}reference{-}reads}\NormalTok{ ref{-}seqs.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}reference{-}taxonomy}\NormalTok{ ref{-}taxonomy.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}classifier}\NormalTok{ classifier.qza}

\CommentTok{\# test the classifier}
\ExtensionTok{qiime}\NormalTok{ feature{-}classifier classify{-}sklearn }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}classifier}\NormalTok{ classifier.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}reads}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc/rep{-}seqs{-}nonchimeric{-}w{-}borderline{-}97{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}classification}\NormalTok{ taxonomy{-}97{-}w{-}borderline{-}no{-}trunc{-}no{-}minmax.qza}

\ExtensionTok{qiime}\NormalTok{ metadata tabulate }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}m{-}input{-}file}\NormalTok{ taxonomy{-}97{-}w{-}borderline{-}no{-}trunc{-}no{-}minmax.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}visualization}\NormalTok{ taxonomy{-}97{-}w{-}borderline{-}no{-}trunc{-}no{-}minmax.qzv}
\end{Highlighting}
\end{Shaded}

\begin{itemize}
\tightlist
\item
  Generate taxa bar plots
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# taxa bar plots}
\ExtensionTok{qiime}\NormalTok{ taxa barplot }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}table}\NormalTok{ uchime{-}dn{-}out{-}97{-}no{-}trunc/table{-}nonchimeric{-}w{-}borderline{-}97{-}no{-}trunc.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}i{-}taxonomy}\NormalTok{ taxonomy{-}97{-}w{-}borderline{-}no{-}trunc{-}no{-}minmax.qza }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}m{-}metadata{-}file}\NormalTok{ filenames.txt }\DataTypeTok{\textbackslash{}}
  \AttributeTok{{-}{-}o{-}visualization}\NormalTok{ taxa{-}bar{-}plots{-}97{-}no{-}trunc{-}w{-}borderline{-}no{-}minmax.qzv}
\end{Highlighting}
\end{Shaded}

\hypertarget{results}{%
\section{Results:}\label{results}}

1 722 650 reads in total Min: 48678; Max: 49499; Mean: 49218.571429

\includegraphics{../../qiime-scripts/biagi-nmr/images/2022-11-14-taxa-barplot.png}

I am too tired to do a clean R version. I'll do it tomorrow.

\end{document}
